doc/developer/design/20230714_optimizer_interface.md

```diff
@@ -0,0 +1,336 @@
+    /// A human-readable exposed internally (useful for debugging).
+    debug_name: String,
+    // Optimizer flags.
+    flags: OptimizerFlags,
+}
+```
+
+The following diagram illustrates the pipeline implemented by `OptimizeMaterializedView`
+
+```mermaid
+flowchart TD
+
+HRE(HirRelationExpr) --> LMP(LocalMirPlan)
+LMP --> GM1("GlobalMirPlan< Unresolved >")
+OMRE(OptimizedMirRelationExpr) --> GM1
+GM1 -.-> GM2("GlobalMirPlan< Resolved >")
+GM2 --> GL(GlobalLirPlan)
+```
+
+Solid edges represent transitions done by `optimize` calls on the various
+`Optimize<From>` implementations on `OptimizeMaterializedView`. The dotted edge
+represents a transition between a non-timestamped `GlobalMirPlan` to a
+timestamped `GlobalMirPlan` that the client can perform after determining the
+execution timestamp for the given statement.
+
+The `LocalMirPlan`, `GlobalMirPlan`, and `GlobalLirPlan` types encapsulate
+intermediate results that wrap the resulting plan and strucutre and associated
+metadata.
+
+The pipeline can be entered both from an `HirRelationExpr` and from an
+`OptimizedMirRelationExpr` — the latter is needed because the
+`Coordinator::bootstrap` code path starts from a catalog item that stores an
+already locally optimized `MirRelationExpr`.
+
+
+## Minimal Viable Prototype
+
+<!--
+Build and share the minimal viable version of your project to validate the
+design, value, and user experience. Depending on the project, your prototype
+might look like:
+
+- A Figma wireframe, or fuller prototype
+- SQL syntax that isn't actually attached to anything on the backend
+- A hacky but working live demo of a solution running on your laptop or in a
+  staging environment
+
+The best prototypes will be validated by Materialize team members as well
+as prospects and customers. If you want help getting your prototype in front
+of external folks, reach out to the Product team in #product.
+
+This step is crucial for de-risking the design as early as possible and a
+prototype is required in most cases. In _some_ cases it can be beneficial to
+get eyes on the initial proposal without a prototype. If you think that
+there is a good reason for skpiping or delaying the prototype, please
+explicitly mention it in this section and provide details on why you you'd
+like to skip or delay it.
+-->
+
+MaterializeInc/materialize#22115 contains an MVP of the design proposed above
+and a safe path towards integrating the new `Optimize` API into the existing
+code base.
+
+The `Optimize` trait itself and its implementations is contained in the
+`mz_adapter::optimize` module. As part of the implementation work, we will also
+move the `mz_adapter::coord::dataflows` module under `mz_adapter::optimize`. In
+its end state, the contents of the `mz_adapter::optimize` module should have a
+minimal set of dependencies against `mz_adapter` so it should be easy to move
+them around and/or create a dedicated `mz_optimizer` crate as part of the
+Platform V2 effort if needed.
```

Comment:
What prevents us from introducing an `mz_optimizer` trait now already? Too much effort to untangle optimizer code from adapter code?